### ISSUE TITLE
[Bug] usage of WithIosKeychainSecurityGroup

### ISSUE THREAD

**Issue by Pabby24** [AzureAD/microsoft-authentication-library-for-dotnet#2614]
**Logs and Network traces**
No relevant errors found in logs.

**Which Version of MSAL are you using ?**

MSAL 2.6.2

**Platform**

xamarin iOS

**What authentication flow has the issue?**
* Desktop / Mobile
    * [ ] Interactive
    * [ ] Integrated Windows Auth
    * [ ] Username Password
    * [ ] Device code flow (browserless)
* Web App 
    * [ ] Authorization code
    * [ ] OBO
* Daemon App 
    * [ ] Service to Service calls

Mobile: Username/Password

Other? - please describe;

**Is this a new or existing app?**

The app is in production. Trying to add MSAL for login using Azure and Facebook.

**Repro**

```csharp
//Login
            AuthenticationClient = PublicClientApplicationBuilder.Create(Constants.ClientId)
            .WithIosKeychainSecurityGroup(Constants.IosKeychainSecurityGroups)
            //.WithIosKeychainSecurityGroup("*")
            .WithB2CAuthority(Constants.AuthoritySignin)
            .WithRedirectUri($"msal{Constants.ClientId}://auth")
            //.WithRedirectUri($"https://jwt.ms")
            .Build();
```

**Expected behavior**
Login using Azure AD or Facebook works as expected.

**Actual behavior**
Login only works in DEV when using .WithIosKeychainSecurityGroup("*"). This does not work when deployed to the app store. If changed to .WithIosKeychainSecurityGroup("com.microsoft.adalcache") or with my identifier, error occurs:

The application does not have keychain access enabled in Entitlements =.plist.

**Possible Solution**

.WithIosKeychainSecurityGroup("*").

**Additional context/ Logs / Screenshots**
Add any other context about the problem here, such as logs and screebshots.

**Comment by jmprieur** [comment]
@Pabby24, before anything, I'd suggest to update to the latest version of MSAL.NET. A lot was fixed since.

**Comment by Pabby24** [comment]
I was not even sure what to put for that as the version as I did not have any MSAL nuget packages installed and it works with the * passed in. I downloaded Microsoft.Identity.Client.Extensions.Msal 2.18.4. I still get the same error. Please advise.

**Comment by Pabby24** [comment]
researched further and installed Microsoft.Identity.Client package. Still same issue.

**Comment by Pabby24** [comment]
FYI, Microsoft.Identity.Client version installed is v4.30.1

**Comment by jmprieur** [comment]
@Pabby24 : did you read this article: https://docs.microsoft.com/azure/active-directory/develop/msal-net-xamarin-ios-considerations#enable-token-cache-sharing-across-ios-applications ?

**Comment by Pabby24** [comment]
Yes, thank you for sending. That is the article I used to initially set up my project to use Microsoft Authentication. That was also the reference I used when trying the fixes I mentioned above installing Microsoft.Identity.Client.Extensions.Msal 2.18.4 and Microsoft.Identity.Client 4.30.1. Results are the same after installing those. I researched more and also tried different values in my entitlements.plist for keychain-access-groups, including using my bundle identifier instead of "com.microsoft.adalcache" in the file and passed into iosKeychainSecurityGroup, and hard coding the app prefix instead of $(AppIdentifierPrefix). It still only works if I use .WithIosKeychainSecurityGroup("*"), using the "*" instead of com.microsoft.adalcache or the bundle identifier, but this does not work when I distribute to the app store.

I will look at the article again, but I have gone step by step several times and am pretty sure I have covered everything listed.

Entitlements line:

<key>keychain-access-groups</key>
	<array>
		<string>$(AppIdentifierPrefix)com.microsoft.adalcache</string>
	</array>

**Comment by Pabby24** [comment]
The entitlements file is definitely referenced in the bundle signing as an additional resource. Xcode is installed in the mac build host.  Entitlements are correct in the WithIosKeychainSecurityGroup and the Entitlements file.

Should I log a ticket with Microsoft. Why does it work with .WithIosKeychainSecurityGroup("*"), but I change to .WithIosKeychainSecurityGroup("com.microsoft.adalcache") and get the error? Note, it only works with the * in debug mode, but in Release mode, I get the same error as if I try in Debug without using *. Not sure what makes it work with wildcard.

**Comment by Pabby24** [comment]
[using Xamarin iOS with MSAL.docx](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/files/6468892/using.Xamarin.iOS.with.MSAL.docx)

I attached a document with code from all applicable classes. Please let me know if this helps or if I can provide anything else. I don't think it is a connection issue to the mac host as I can connect fine and I can build/archive and deploy to the app store just fine, but the actual authentication to MSAL only works when I have the * in the call to WithIosKeychainSecurityGroup, and it only works in Debug mode, not release mode.

**Comment by Pabby24** [comment]
here is a screenshot of the error when  I try to authenticate (facebook or Azure):

![Keychain Error](https://user-images.githubusercontent.com/83847208/118046745-bffcd580-b347-11eb-8ff0-9772014d2977.PNG)

**Comment by Pabby24** [comment]
I logged a ticket within VS2019 to Microsoft and they directed me back to here for guidance. Please let me know if you have any updates. I have attached my code snippets and my connection to mac/XCode seem fine and my Entitlements file is definitely referenced in the bundle signing as an additional resource. As I noted, everything works fine when I use *, but that is not allowed when deploying to the app store.

**Comment by Pabby24** [comment]
One other note, I don't need to be connected to the mac build host to test this, but the behavior is the same connected or not. Let me know if that is correct or if there is something specific to do while connected to the build host.

**Comment by Pabby24** [comment]
Another note: If I try to test on a simulator on the remote host mac, I can get it to work using com.microsoft.adalcache passed into .WithIosKeychainSecurityGroup and entered in Entitlements. The issue as actually when you try it on a device via debug or deployed to the store (and downloading via Testflight).

So why would the keychain access group for using MSAL work on simulator but not a device?

**Comment by Pabby24** [comment]
Any updates? This has been very frustrating to try to get authentication for users. I can try to find an alternate solution if needed, but would like to have this work if possible.

**Comment by bgavrilMS** [comment]
@Pabby24 - [here's ](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/fac5d52f802de1c523c734b349a8e39c7d65a9f3/src/client/Microsoft.Identity.Client/Platforms/iOS/iOSTokenCacheAccessor.cs#L211) where MSAL tries to save some tokens to the KeyChain. 

Can you do that directly from your app, i.e. save some text to the KeyChain ? You can use any text for Account, Sevice and Generic and the number `2001` for CreatorType.

**Comment by Pabby24** [comment]
I was able to get it to work over the weekend and deploy to the store and working in testflight. I kept the wildcard * to pass into iosKeychainSecurityGroup and changed a couple of things:

1 - Added * as a keychain access group in Entitlements file
2 - Changed Bundle signing in project properties to Manual Provisioning instead of Auto

I think #2 may have fixed it as I tried debugging in Release Configuration and noticed VS studio was still using the DEV Provisioning Profile in Release Config.

Hope this helps anyone else with the issue and thanks for the responses!
